src/MyPlot/EventListener.php

```diff
@@ -234,56 +234,57 @@ public function onEntityMotion(EntityMotionEvent $event) {
 			return;
 		}
-		if($event->getFrom()->floor()->equals($event->getTo()->floor())) return; // Save hell-a-lot of performance by ignoring movements on the same block, for example rotating the head
 
 		$levelName = $event->getPlayer()->getLevel()->getName();
 		if (!$this->plugin->isLevelLoaded($levelName))
 			return;
 
 		$plot = $this->plugin->getPlotByPosition($event->getTo());
 		$plotFrom = $this->plugin->getPlotByPosition($event->getFrom());
-		if ($plot !== null and $plot !== $plotFrom) {
+		if($plot instanceof Plot and $plotFrom === null) {
 			$this->plugin->getServer()->getPluginManager()->callEvent($ev = new MyPlotPlayerEnterPlotEvent($this->plugin, $event->getPlayer()->getName(), $plot, $event->getPlayer()));
 			if($ev->isCancelled()) {
 				$event->setCancelled();
 				return;
 			}
+		} elseif($plotFrom instanceof Plot and $plot === null) {
+			$this->plugin->getServer()->getPluginManager()->callEvent($ev = new MyPlotPlayerLeavePlotEvent($this->plugin, $event->getPlayer()->getName(), $plot, $event->getPlayer()));
+			if($ev->isCancelled()) {
+				$event->setCancelled();
+				return;
+			}
+		}
 
+		if($plot !== null) {
 			if($plot->isDenied($event->getPlayer()->getName())) {
 				$event->setCancelled();
 				return;
 			}
 
 			$plotName = TextFormat::GREEN . $plot;
-			$popup = $this->plugin->getLanguage()->translateString("popup", [$plotName]);
-			if(strpos($plot,"-0")) { // Jason, tell me WTF this is - strpos needs a string. $plot is not a string.
+			$popup    = $this->plugin->getLanguage()->translateString("popup", [$plotName]);
+			if(strpos($plot, "-0")) {
 				return;
 			}
 			if ($plot->owner != "") {
-				$owner = TextFormat::GREEN . $plot->owner;
-				$ownerPopup = $this->plugin->getLanguage()->translateString("popup.owner", [$owner]);
-				$paddingSize = floor((strlen($popup) - strlen($ownerPopup)) / 2);
-				$paddingPopup = str_repeat(" ", max(0, -$paddingSize));
+				$owner             = TextFormat::GREEN . $plot->owner;
+				$ownerPopup        = $this->plugin->getLanguage()->translateString("popup.owner", [$owner]);
+				$paddingSize       = floor((strlen($popup) - strlen($ownerPopup)) / 2);
+				$paddingPopup      = str_repeat(" ", max(0, -$paddingSize));
 				$paddingOwnerPopup = str_repeat(" ", max(0, $paddingSize));
-				$popup = TextFormat::WHITE . $paddingPopup . $popup . "\n" .
-					TextFormat::WHITE . $paddingOwnerPopup . $ownerPopup;
+				$popup             = TextFormat::WHITE . $paddingPopup . $popup . "\n" . TextFormat::WHITE . $paddingOwnerPopup . $ownerPopup;
 			} else {
-				$ownerPopup = $this->plugin->getLanguage()->translateString("popup.available");
-				$paddingSize = floor((strlen($popup) - strlen($ownerPopup)) / 2);
-				$paddingPopup = str_repeat(" ", max(0, -$paddingSize));
+				$ownerPopup        = $this->plugin->getLanguage()->translateString("popup.available");
+				$paddingSize       = floor((strlen($popup) - strlen($ownerPopup)) / 2);
```

Comment:
neat formatting

Reply:
Phpstorm did that on it’s own 🤦‍♂️

Reply:
Totally, my PHPStorm _always_ does that too